### ISSUE TITLE
GUI install under OS X: handling of framework-python

### ISSUE THREAD

**Issue by cjayb** [psychopy/psychopy#1326]
TL;DR If you want to run the `psychopyApp.py` GUI on an OS X dev-install, call

```bash
python.app `which psychopyApp.py`  # or equivalently use: pythonw
```

## Description

Even on successful OS X compilations, and after this locale-fix #1325, the distutils install leaves the script `psychopyApp.py` with the shebang

```bash
#!/Users/cjayb/miniconda_envs/psychopy/bin/python
```

_i.e._, one that is not compiled against the "Framework" libs and therefore does not get access to the screen:

```bash
This program needs access to the screen.
Please run with a Framework build of python, and only when you are
logged in on the main display of your Mac.
```

Both `pythonw` and `python.app` are [easily installed using conda](https://gist.github.com/cjayb/67d8f6ada85f8fcd58ff1d4e1fc20764), and it's possible to specify the interpreter for `scripts`:

```bash
python setup.py build --executable=`which python.app`
```

I've also tried simply:

```bash
python.app setup.py install --user
```

but I can't get the pieces to work together! Whatever I try, the shebang remains just `python`.

Any ideas on whether this can be fixed in `setup.py`?

**Comment by peircej** [comment]
I don't really understand the aim here. You're trying to construct the app bundle from a conda installation? Or you're tryng to *avoid* a conda python getting involved?
To be fair though I don't know the answer as I use regular python.org install and pip libs.

**Comment by cjayb** [comment]
This is just a minor inconvenience, and a consequence of using the conda-packaged python over the vanilla .org, which compiles against OS X frameworks and thus doesn't suffer from `wx` lacking access to the screen. No need to complicate `setup.py` for this!

Feel free to close this. If you like, you can add the very first line of code to your dev-docs, in case conda-users find it useful:

```bash
python.app `which psychopyApp.py`
```

**Comment by hoechenberger** [comment]
Anaconda is very common among scientists and I recommend it to all my colleagues and students. I can reproduce @cjayb's report:

```zsh
[hoechenberger:~]$ source activate new_psychopy
(new_psychopy) [hoechenberger:~]$ which python
/Users/hoechenberger/anaconda3/envs/new_psychopy/bin/python
(new_psychopy) [hoechenberger:~]$ python --version
Python 2.7.13 :: Anaconda 4.3.0 (x86_64)
(new_psychopy) [hoechenberger:~]$ python ~/Development/psychopy/psychopy/app/psychopyApp.py
This program needs access to the screen.
Please run with a Framework build of python, and only when you are
logged in on the main display of your Mac.
(new_psychopy) [hoechenberger:~]$ pythonw ~/Development/psychopy/psychopy/app/psychopyApp.py
2017-03-22 13:14:07.133 python[6020:537744] ApplePersistenceIgnoreState: Existing state will not be touched. New state will be written to /var/folders/75/zwcv5c8j5sz3c1_6m60xpct80000gp/T/com.continuum.python.savedState
^C
(new_psychopy) [hoechenberger:~]$
```
I think `setup.py` should check if the current platform is macOS ("Darwin") and use `pythonw` in that case.

**Comment by hoechenberger** [comment]
By "use", I mean add `pythonw` to the shebang line instead of `python`. Also I wonder if the interpreter specified there should not by "location-agnostic", i.e., `#!/usr/bin/env pythonw2.7`.

**Comment by peircej** [comment]
I'm happy to see changes to setup.py if those will help (and assuming they won't break other things). I don't think I've ever made use of the shebang line. Always specify the python that will launch the app. My `.bash_profile` has this (and obviously pythonw would be easy to switch in)

`alias psp='python $HOME/code/psychopy/git/psychopy/app/psychopyApp.py'`

and then I launch the app with `psp`. Maybe that's why I haven't really understood what the issue is here. But, as I say, if there's something that would be useful and won't mess with anyone else then I'm happy for it to be done. (There have been a number of changes to the shebang line, always by other people)

**Comment by hoechenberger** [comment]
I also wonder whether the line should not be "location-agnostic", i.e., `#!/usr/bin/env python(w)2` instead of specifying an exact path? Could have undesired side-effects in realistic situations?

x-ref:
https://github.com/ContinuumIO/anaconda-issues/issues/199

**Comment by hoechenberger** [comment]
On Unix-like platforms, couldn't we simply create a symlink to `psychopy/psychopy/app/psychopyApp.py`? Or is this not supported by distutils?

**Comment by hoechenberger** [comment]
Ah no, scratch that. The symlink would require the installer to alter the original file `psychopy/psychopy/app/psychopyApp.py` to use `pythonw` in the shebang line. Argh! So we should definitely stick to some sort of "separate invocation script", which the setup process would adjust such that it would run `psychopyApp.py` with the correct interpreter.

**Comment by hoechenberger** [comment]
Thinking some more about this, I'm starting to believe this is actually a bug in Anaconda that we shouldn't have to deal with. I would propose to update the install documentation and wait for the Anaconda people to fix this already. There's many people affected by this odd behavior, not just users of PsychoPy...

**Comment by hoechenberger** [comment]
...or setuptools should work around it:
https://github.com/pypa/setuptools/issues/410

**Comment by djmannion** [comment]
Related to the comment 5 posts above about a "location-agnostic" path: currently, installing from pip (the wheel) on linux causes the first line on psychopyApp.py to be `#!/Library/Frameworks/Python.framework/Versions/2.7/Resources/Python.app/Contents/MacOS/Python`. A location-agnostic path would presumably also fix that.

**Comment by hoechenberger** [comment]
The problem is that setuptools automatically changes the shebang line:
https://docs.python.org/2/distutils/setupscript.html#installing-scripts

**Comment by peircej** [comment]
OK, in the repository the shebang is `#!/usr/bin/env python2`
This was being changed by setuptools during build of eggs and wheels to the path on my local machine (dumbest idea ever)
Eggs installed with easy_install actually then get changed *again* during install so those were fine but pip doesn't touch the shebang

BUT, this was using setuptools 28.3.0. I've just updated setuptools to 34.3.2 on my build machine and it seems not to touch shebnag anymore for script copying so we'll stick with `#!/usr/bin/env python2`

Hopefully that will fix for everyone. In the meantime you can:

- install the source install (zip file) instead of the egg/wheel
- install with easy_install
- avoid using the shebang line altogether by specifying your python executable when launching the script

**Comment by peircej** [comment]
Oh, by the way, I'm happy for the shebang to be changed to pythonw2 for the psyhopyApp.py if you think that will work everywhere

**Comment by hoechenberger** [comment]
> Oh, by the way, I'm happy for the shebang to be changed to pythonw2 for the psyhopyApp.py if you think that will work everywhere

I'm afraid `pythonw` is not universally available on all platforms.
